### PR TITLE
Update Scheduler Pallet Documentation

### LINKED ISSUE #14715: Document safety constraints of `pallet-scheduler`
There is a lot of insider information, spread mostly in forum posts like https://forum.polkadot.network/t/pallet-idea-safe-scheduler/1009 that the existing scheduler pallet is unsafe, and should only be used internally by the runtime. In other words, scheduling operations **should not be allowed** by arbitrary external users. 

I am worried that, similar to offchain workers, this pallet it as first glance a silver bullet that might lead to someone designing an application with the wrong assumption, thinking that this pallet is going magically schedule things for them across blocks, while it won't, and it can easily lead a parachain to stall. 

This information is vital and should be in the source docs of the pallet, not the heads of a few. 

Additionally, a warning about the change of decoding for stored calls is also a good point to be mentioned. 

Update https://paritytech.github.io/substrate/master/pallet_scheduler/index.html to be aligned with the house [pallet documentation guidelines](https://github.com/paritytech/substrate/blob/master/docs/DOCUMENTATION_GUIDELINES.md), and make sure this point is also covered.

### PR DESCRIPTION
**Pull request by snowmead**
# Description

Updates the documentation of the [Scheduler Pallet](https://github.com/paritytech/substrate/tree/master/frame/scheduler) based on pallet documentation guidelines. 

_Warning_ section added to inform users of some anti-patterns when scheduling runtime calls due to misconceptions of the pallet.

Closes #14715

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3355483

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3355482

**Comment by muharem** [comment]
there is a new documentation guidelines you might wanna check https://github.com/paritytech/substrate/blob/master/docs/DOCUMENTATION_GUIDELINES.md

File: frame/scheduler/Cargo.toml [hunk 1]
```diff
@@ -10,16 +10,21 @@ description = "FRAME Scheduler pallet"
...
 [dependencies]
-codec = { package = "parity-scale-codec", version = "3.6.1", default-features = false, features = ["derive"] }
+codec = { package = "parity-scale-codec", version = "3.6.1", default-features = false, features = [
+	"derive",
+] }
 log = { version = "0.4.17", default-features = false }
-scale-info = { version = "2.5.0", default-features = false, features = ["derive"] }
+scale-info = { version = "2.5.0", default-features = false, features = [
+	"derive",
+] }
```
Comment (kilogold) [inline frame/scheduler/Cargo.toml:19]: Cosmetic:
Breaks consistency with the rest of the file.

File: frame/scheduler/src/lib.rs [hunk 2]
```diff
@@ -15,38 +15,63 @@
...
-//! specified block number or at a specified period. These scheduled dispatches
+//! This Pallet exposes capabilities for scheduling runtime calls to occur at a
+//! specified block number or at a specified period. These scheduled runtime calls
 //! may be named or anonymous and may be canceled.
 //!
-//! **NOTE:** The scheduled calls will be dispatched with the default filter
+//! **NOTE:** The scheduled runtime calls will be dispatched with the default filter
 //! for the origin: namely `frame_system::Config::BaseCallFilter` for all origin
 //! except root which will get no filter. And not the filter contained in origin
 //! use to call `fn schedule`.
```
Comment (kilogold) [inline frame/scheduler/src/lib.rs:31]:
```suggestion
//! **NOTE:** Instead of the filter contained in the origin used to call `fn schedule`, scheduled runtime calls will be 
//! dispatched with the default filter for the origin: namely `frame_system::Config::BaseCallFilter` for all origin
//! types (except root which will get no filter).
```

**Review by kilogold — COMMENTED** [review]
I like what I'm seeing here.

File: frame/scheduler/src/lib.rs [hunk 3]
```diff
@@ -15,38 +15,62 @@
...
+//! 	Use [`frame_support::traits::Hooks::on_idle`] hook for runtime calls that can be executed
+//!   after some block but do not need to be executed at a specific block.
+//! * [ ] No unbounded iterations.
+//! * [ ] Runtime call weight exceeds configured [`Config::MaximumWeight`].
+//! 	If you are scheduling a heavy weight runtime call, make sure to set the `MaximumWeight` type
+//! according to your chain's needs. You can find the weight conversions to real time from these
+//! [`sp_weights::constants`].
+//!
+//! Please be aware that any scheduled runtime calls executed in a future block may __fail__ or may
+//! result in __unintended behavior__ since the runtime could have upgraded between the time of
```
Comment (KiChjang) [inline frame/scheduler/src/lib.rs:57]:
```suggestion
//! result in __undefined behavior__ since the runtime could have upgraded between the time of
```

Sounds like this would really nail the point home.

File: frame/scheduler/src/lib.rs [hunk 4]
```diff
@@ -15,38 +15,62 @@
...
+//! scheduling and execution. For example, the runtime upgrade could have:
+//!
+//! * Modified the implementation of the runtime call (runtime specification upgrade).
+//!     * Could lead to unintended behavior.
```
Comment (KiChjang) [inline frame/scheduler/src/lib.rs:61]:
```suggestion
//!     * Could lead to undefined behavior.
```

File: frame/scheduler/src/lib.rs [hunk 5]
```diff
@@ -15,38 +15,62 @@
...
+//! * Removed or changed the ordering/index of the runtime call.
+//!     * Could fail due to the runtime call index not being part of the `Call`.
+//!     * Could lead to unintended behavior, such as executing another runtime call with the same
```
Comment (KiChjang) [inline frame/scheduler/src/lib.rs:64]:
```suggestion
//!     * Could lead to undefined behavior, such as executing another runtime call with the same
```

File: frame/scheduler/src/lib.rs (hunk 3 above)
Comment (snowmead) [reply frame/scheduler/src/lib.rs:57]: Agreed.

File: frame/scheduler/src/tests.rs [hunk 6]
```diff
@@ -30,6 +30,7 @@ use sp_runtime::traits::Hash;
 use substrate_test_utils::assert_eq_uvec;
 
 #[test]
+#[docify::export]
```
Comment (kianenigma) [inline frame/scheduler/src/tests.rs:33]: You can try and add a bit of explanation/docs to these tests as well, if suitable.

File: frame/scheduler/src/lib.rs [hunk 7]
```diff
@@ -15,38 +15,62 @@
...
-//! * `cancel_named` - the named complement to the cancel function.
-
+//! ### Warning
+//!
+//! This Pallet executes all scheduled runtime calls in the
+//! [`frame_support::traits::Hooks::on_initialize`] hook.
+//!
+//! **Do NOT schedule runtime calls if they satisfy any of the following points:**
+//!
+//! * [ ] Runtime call does not need to be executed at a specific block (not a hard requirement).
```
Comment (kianenigma) [inline frame/scheduler/src/lib.rs:47]: I think what you saying here is more a property of `on_initialize` than this pallet. I would phrase is as this sequence: 

1. This pallet uses `on_initialize`
2. `on_initialize` has a mandatory property to it (which you can make sure is documented in the `Hooks` trait), and should only be used for _really_ mandatory stuff. There, you can also mention `on_idle`
4. Therefore, this pallet is also suitable for scheduling _really mandatory_ stuff. 


In other words, I don't think scheduler and `on_idle` are apples to apples.

File: frame/scheduler/src/lib.rs [hunk 8]
```diff
@@ -15,38 +15,62 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//! # Scheduler
-//! A Pallet for scheduling dispatches.
+//! # Scheduler Pallet
```
Comment (kianenigma) [inline frame/scheduler/src/lib.rs:18]: Can you add the fancy heading as well please, from the pallet docs template?

File: frame/scheduler/src/lib.rs [hunk 9]
```diff
@@ -15,38 +15,62 @@
...
-//! for the origin: namely `frame_system::Config::BaseCallFilter` for all origin
-//! except root which will get no filter. And not the filter contained in origin
-//! use to call `fn schedule`.
+//! **NOTE:** Instead of the filter contained in the origin used to call `fn schedule`, scheduled runtime calls will be 
+//! dispatched with the default filter for the origin: namely `frame_system::Config::BaseCallFilter` for all origin
+//! types (except root which will get no filter).
 //!
 //! If a call is scheduled using proxy or whatever mecanism which adds filter,
-//! then those filter will not be used when dispatching the schedule call.
+//! then those filter will not be used when dispatching the schedule runtime call.
```
Comment (kianenigma) [inline frame/scheduler/src/lib.rs:33]: Line wrapping seems too short here.

Comment (kianenigma) [inline frame/scheduler/src/lib.rs:33]: Line wrapping seems too short here.

File: frame/scheduler/src/lib.rs [hunk 10]
```diff
@@ -15,38 +15,62 @@
...
+//! including its configuration trait, dispatchables, storage items, events and errors.
 //!
-//! * `schedule` - schedule a dispatch, which may be periodic, to occur at a specified block and
-//!   with a specified priority.
-//! * `cancel` - cancel a scheduled dispatch, specified by block number and index.
-//! * `schedule_named` - augments the `schedule` interface with an additional `Vec<u8>` parameter
-//!   that can be used for identification.
-//! * `cancel_named` - the named complement to the cancel function.
-
+//! ### Warning
```
Comment (kianenigma) [inline frame/scheduler/src/lib.rs:40]: Warnings and Examples are level 2 heading afair, not 3.

**Review by kianenigma — APPROVED** [review]
Some suggestions, but all in all looks good!

**Review by sam0x17 — APPROVED** [review]
few grammar / wording nits, then looks good

File: frame/scheduler/src/lib.rs [hunk 11]
```diff
@@ -15,40 +15,64 @@
...
-//!   with a specified priority.
-//! * `cancel` - cancel a scheduled dispatch, specified by block number and index.
-//! * `schedule_named` - augments the `schedule` interface with an additional `Vec<u8>` parameter
-//!   that can be used for identification.
-//! * `cancel_named` - the named complement to the cancel function.
-
+//! 2. Scheduling a preimage hash of a runtime call at a specifc block
+#![doc = docify::embed!("src/tests.rs", scheduling_with_preimages_works)]
 // Ensure we're `no_std` when compiling for Wasm.
 #![cfg_attr(not(feature = "std"), no_std)]
```
Comment (muharem) [inline frame/scheduler/src/lib.rs:52]: lets move this after the docs block

File: frame/scheduler/src/tests.rs [hunk 12]
```diff
@@ -30,45 +30,72 @@ use sp_runtime::traits::Hash;
...
 	new_test_ext().execute_with(|| {
+		// Call to schedule
 		let call =
 			RuntimeCall::Logger(LoggerCall::log { i: 42, weight: Weight::from_parts(10, 0) });
+
+		// BaseCallFilter should be implemented to accept `Logger::log` runtime call which is
+		// implemented for `BaseFilter` in the mock runtime
 		assert!(!<Test as frame_system::Config>::BaseCallFilter::contains(&call));
+
+		// Schedule call to be executed at the the 4th block
```
Comment (naterarmstrong) [inline frame/scheduler/src/tests.rs:44]:
```suggestion
		// Schedule call to be executed at the 4th block
```

File: frame/scheduler/src/tests.rs [hunk 13]
```diff
@@ -30,45 +30,72 @@ use sp_runtime::traits::Hash;
...
+		// Register preimage on chain
 		assert_ok!(Preimage::note_preimage(RuntimeOrigin::signed(0), call.encode()));
 		assert!(Preimage::is_requested(&hash));
+
+		// `log` runtime call should not have executed yet
 		run_to_block(3);
 		assert!(logger::log().is_empty());
+
 		run_to_block(4);
+		// preimage should no have been removed when executed by the scheduler
```
Comment (naterarmstrong) [inline frame/scheduler/src/tests.rs:93]:
```suggestion
		// preimage should not have been removed when executed by the scheduler
```

**Review by naterarmstrong — APPROVED** [review]
❤️

**Review by KiChjang — APPROVED** [review]

**Review by wentelteefje — APPROVED** [review]

**Review by gilescope — APPROVED** [review]

**Comment by gilescope** [comment]
bot merge
